content/posts/notes-on-2pc.md

```diff
@@ -5,38 +5,82 @@ date: 2019-09-04
 draft: true
 ---
 
-If there's one protocol every distributed system's engineer knows it's probably Two-Phase Commit (2PC). That said the understanding often doesn't match the popularity. In this post will discuss misconceptions around 2PC that we have encountered. This is not yet another description of 2PC (TM) - if you need a refresher you might want to read [this](link-to-2pc-description).
+If there's a distributed protocol every software engineer knows it's Two-Phase Commit also know as 2PC. Although, in use for several decades[^1], it's been in steady decline mainly due to lack of support in cloud environments. 
 
-### 2PC == MS DTC
-MS DTC (or XA if you come from the Java space) is an implementation of 2PC - one of many possible. Among others, it comes with concepts like `TransactionManagers` and `ResourceManagers`, implements security between parts of the system and provides tooling to figure out what's going on in your system. As we will soon see it's important to make it clear if we talk about MS DTC or 2PC.
+For quite a long time it has been a de-facto standard for building enterprise distributed systems. That said, with the cloud becoming the default deployment model designers need to learn how to build reliable systems without. 
 
-### 2PC applies to databases only
-2PC tells nothing about resources participating in the protocol. In MS DTC those could be databases, queueing systems, Web Services etc.
+Answering the question of how 2PC can be replaced requires an understanding of what it was, that the protocol provided in the first place. In spite of its popularity, there are plenty of misconceptions around 2PC. This post aims to clarify at least some of these.      
 
-### 2PC provides atomic visibility
-2PC is an atomic commit protocol meaning that all participants will eventually commit. It **does not** provide atomic visibility i.e. making sure that it's impossible to see partial commit results.  
+NOTE: This is not "yet another introduction to 2PC"(TM). If you need a refresher read one of many descriptions out there before continuing.
 
-Two databases in snapshot isolation mode and reads from one and the other. 
+### 2PC does not provide "transactions"
+2PC is an atomic commit protocol meaning all participants will **eventually commit** if all voted "YES" or leave the system unchanged otherwise. When a commit operation triggered by the user finishes, either all local modifications have been applied or none of them. The commit can take arbitrarily long to complete. In some failure scenarios, it will hang forever.
```

Comment:
```suggestion
2PC is an atomic commit protocol meaning all participants will **eventually commit** if all voted "YES" or leave the system unchanged otherwise. When a commit operation triggered by the user finishes, either all local modifications have been applied or none of them has. The commit can take arbitrarily long to complete. In some failure scenarios, it will hang forever.
```